iree/compiler/Codegen/SPIRV/SPIRVExpandIfRegion.cpp

```diff
@@ -0,0 +1,257 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns and passes for expanding scf.if's regions
+// by pulling in ops before and after the scf.if op into both regions of the
+// scf.if op.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Affine/Analysis/AffineStructures.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/SCF/Passes.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Dialect/SCF/Utils/AffineCanonicalizationUtils.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-spirv-expand-if-region"
+
+// Note: A pull request is open to upstream this pattern:
+//   https://reviews.llvm.org/D117019
+// After it lands, this pattern can be replaced.
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Pulls ops at the same nest level as the given `ifOp` into both regions of
+/// the if `ifOp`.
+static FailureOr<scf::IfOp> pullOpsIntoIfRegions(
+    scf::IfOp ifOp, function_ref<bool(Operation *)> canMoveToRegion,
+    RewriterBase &rewriter) {
+  // Need to pull ops into both regions.
+  if (!ifOp.elseBlock()) return failure();
+
+  // Expect to only have one block in the enclosing region. This is the common
+  // case for the level where we have structured control flows and it avoids
+  // traditional control flow and simplifies the analysis.
+  if (!llvm::hasSingleElement(ifOp->getParentRegion()->getBlocks()))
+    return failure();
+
+  SmallVector<Operation *> allOps;
+  for (Operation &op : ifOp->getBlock()->without_terminator())
+    allOps.push_back(&op);
+
+  // If no ops before or after the if op, there is nothing to do.
+  if (allOps.size() == 1) return failure();
+
+  // Return true if the given `op` is in the same region as the scf.if op.
+  auto isFromSameRegion = [ifOp](Operation *op) {
+    return op->getParentRegion() == ifOp->getParentRegion();
+  };
+
+  // Collect ops before and after the scf.if op.
+  auto allPrevOps = llvm::makeArrayRef(allOps).take_while(
+      [&ifOp](Operation *op) { return op != ifOp.getOperation(); });
+  auto allNextOps =
+      llvm::makeArrayRef(allOps).drop_front(allPrevOps.size() + 1);
+
+  // Find previous ops that cannot be moved into the regions.
+  SetVector<Operation *> stickyPrevOps;
+  // We cannot move the op into the region if
+  // - The op is part of the backward slice for computing conditions.
```

Comment:
More to the point above. These kind of analysis are what i call "high entropy patterns", i.e. a lot of things need to fall into place for these to work. The conditions here for example prevent these from being small composable patterns for scalar optimizations. Looking more here, maybe the structured control flow here is the bad place to do such optimizations. These should fall out of standard peep hole optimizations on traditional cfgs.

Reply:
Totally understood your points. I agree that we want to strive to write small and composable patterns, but there are cases that due to the fundamental nature of the op, it's hard to make it very small. Control flow like ops are in this category. Or basically anything that fiddles with regions. They won't be as simple as other folders/canonicalization pattern. I also thought about how to "bypass" this but didn't find an easy way. I need to see the tiled and fused `tensor.pad` to be able to create the conditions for the `scf.if` op switching between the fast and slow path. So it need to happen after tiling and fusion.